### PR TITLE
fix: fix index bug and add test to check it

### PR DESCRIPTION
**Pull request by mustafasrepo**
# Which issue does this PR close?

Closes [#7118](https://github.com/apache/arrow-datafusion/issues/7118).

# Rationale for this change
Thanks @alamb and @crepererum for filing the issue.

# What changes are included in this PR?

# Are these changes tested?

# Are there any user-facing changes?

### CODE REVIEW

**Review by jackwener — APPROVED** [review]

**Comment by jonahgao** [comment]
Hi, @mustafasrepo  
Where is the variable `orderings` used? I seem to only see updates to it, or maybe I missed something.
https://github.com/apache/arrow-datafusion/blob/350d4ca8a895a53f51935741ee35e5b79120829c/datafusion/physical-expr/src/aggregate/first_last.rs#L161

**Comment by mustafasrepo** [comment]
> Hi, @mustafasrepo Where is the variable `orderings` used? I seem to only see updates to it, or maybe I missed something.
> 
> https://github.com/apache/arrow-datafusion/blob/350d4ca8a895a53f51935741ee35e5b79120829c/datafusion/physical-expr/src/aggregate/first_last.rs#L161

`Aggregate::Final` or `Aggregate::FinalPartitioned` can require its ordering according to values inside `orderings`. Hence, these stages can get their values with appropriate ordering. However, I will change the implementation. Hopefully, in the new design, usage of it will be much more explicit. And `merge_batch` implementation will not rely on outside mechanism during merging.

**Comment by jonahgao** [comment]
@mustafasrepo  Thank you for explaining  👍

File: datafusion/core/tests/sqllogictests/test_files/groupby.slt [hunk 1]
```diff
@@ -3348,3 +3348,53 @@ ORDER BY l.sn
...
+Aggregate: groupBy=[[]], aggr=[[LAST_VALUE(foo.x)]]
+--TableScan: foo projection=[x]
+physical_plan
+AggregateExec: mode=Final, gby=[], aggr=[LAST_VALUE(foo.x)]
+--CoalescePartitionsExec
+----AggregateExec: mode=Partial, gby=[], aggr=[LAST_VALUE(foo.x)]
+------MemoryExec: partitions=8, partition_sizes=[1, 0, 0, 0, 0, 0, 0, 0]
+
+query I
+SELECT FIRST_VALUE(x)
```
Comment (alamb) [inline datafusion/core/tests/sqllogictests/test_files/groupby.slt:3382]: I recommend also testing `SELECT LAST_VALUE(x) GROUP BY x` (or maybe add another column and `GROUP BY y`) as the grouping operator is different in that case so the code flow might be different

**Review by alamb — APPROVED** [review]
Thank you @mustafasrepo  -- lightning fast turnaround

File: datafusion/core/tests/sqllogictests/test_files/groupby.slt (hunk 1 above)
Comment (mustafasrepo) [reply datafusion/core/tests/sqllogictests/test_files/groupby.slt:3382]: Added these tests

**Comment by alamb** [comment]
Thanks again @mustafasrepo

**Pull request merged**